cmd/argoexec/commands/artifact/delete.go

```diff
@@ -9,40 +9,112 @@ import (
+				}
+
+				artifactGCTaskInterface := workflowInterface.ArgoprojV1alpha1().WorkflowArtifactGCTasks(namespace)
+				labelSelector := fmt.Sprintf("%s = %s", common.LabelKeyArtifactGCPodName, podName)
+
+				taskList, err := artifactGCTaskInterface.List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
+				if err != nil {
+					return err
+				}
 
-			if err := drv.Delete(a); err != nil {
-				return fmt.Errorf("failed to delete artifact: %w", err)
+				err = deleteArtifacts(taskList, cmd, artifactGCTaskInterface)
+				if err != nil {
+					return err
+				}
 			}
 			return nil
 		},
 	}
 }
 
-type resources struct{}
+func deleteArtifacts(taskList *v1alpha1.WorkflowArtifactGCTaskList, cmd *cobra.Command, artifactGCTaskInterface wfv1alpha1.WorkflowArtifactGCTaskInterface) error {
+	for _, task := range taskList.Items {
+		task.Status.ArtifactResultsByNode = make(map[string]v1alpha1.ArtifactResultNodeStatus)
+		for nodeName, artifactNodeSpec := range task.Spec.ArtifactsByNode {
+
+			var archiveLocation *v1alpha1.ArtifactLocation
+			artResultNodeStatus := v1alpha1.ArtifactResultNodeStatus{ArtifactResults: make(map[string]v1alpha1.ArtifactResult)}
+			if artifactNodeSpec.ArchiveLocation != nil {
+				archiveLocation = artifactNodeSpec.ArchiveLocation
+			}
+
+			var resources resources
+			resources.Files = make(map[string][]byte) // same resources for every artifact
+			for _, artifact := range artifactNodeSpec.Artifacts {
+				if archiveLocation != nil {
+					artifact.Relocate(archiveLocation)
+				}
+
+				drv, err := executor.NewDriver(cmd.Context(), &artifact, resources)
+				if err != nil {
+					return err
+				}
+
+				err = drv.Delete(&artifact)
+				if err != nil {
+					errString := err.Error()
+					artResultNodeStatus.ArtifactResults[artifact.Name] = v1alpha1.ArtifactResult{Name: artifact.Name, Success: false, Error: &errString}
+				} else {
+					artResultNodeStatus.ArtifactResults[artifact.Name] = v1alpha1.ArtifactResult{Name: artifact.Name, Success: true, Error: nil}
+				}
+			}
+
+			task.Status.ArtifactResultsByNode[nodeName] = artResultNodeStatus
+		}
+		patch, err := json.Marshal(map[string]interface{}{"status": v1alpha1.ArtifactGCSStatus{ArtifactResultsByNode: task.Status.ArtifactResultsByNode}})
+		if err != nil {
+			return err
+		}
+		_, err = artifactGCTaskInterface.Patch(context.Background(), task.Name, types.MergePatchType, patch, metav1.PatchOptions{}, "status")
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type resources struct {
+	Files map[string][]byte
+}
 
 func (r resources) GetSecret(ctx context.Context, name, key string) (string, error) {
-	file, err := os.ReadFile(filepath.Join(common.SecretVolMountPath, name, key))
-	return string(file), err
 
+	path := filepath.Join(common.SecretVolMountPath, name, key)
+	if file, ok := r.Files[path]; ok {
+		return string(file), nil
+	}
+
+	file, err := os.ReadFile(path)
+	r.Files[path] = file
+	return string(file), err
 }
```

Comment:
This is good except I'm thinking about the case where there's an error reading the file on line 115, but then the error doesn't get returned on line 112 later.
Two options I can think of:
1. don't just map to the []byte but instead to a new struct which includes both the []byte and the err
2. in the case of error, don't bother storing it in the map